qiskit_experiments/library/randomized_benchmarking/clifford_synthesis.py

```diff
@@ -0,0 +1,101 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Clifford synthesis plugins for randomized benchmarking
+"""
+from __future__ import annotations
+
+from typing import Sequence
+
+from qiskit.circuit import QuantumCircuit, Operation
+from qiskit.circuit.equivalence_library import SessionEquivalenceLibrary as sel
+from qiskit.exceptions import QiskitError
+from qiskit.synthesis.clifford import synth_clifford_full
+from qiskit.transpiler import PassManager, CouplingMap, Layout, Target
+from qiskit.transpiler.passes import (
+    SabreSwap,
+    LayoutTransformation,
+    BasisTranslator,
+    CheckGateDirection,
+    GateDirection,
+    Optimize1qGatesDecomposition,
+)
+from qiskit.transpiler.passes.synthesis.plugin import HighLevelSynthesisPlugin
+
+
+class RBDefaultCliffordSynthesis(HighLevelSynthesisPlugin):
+    """Default Clifford synthesis plugin for randomized benchmarking."""
+
+    def run(
+        self,
+        high_level_object: Operation,
+        coupling_map: CouplingMap | None = None,
+        target: Target | None = None,
+        qubits: Sequence | None = None,
+        **options,
+    ) -> QuantumCircuit:
+        """Run synthesis for the given Clifford.
+
+        Args:
+            high_level_object: The operation to synthesize to a
+                :class:`~qiskit.circuit.QuantumCircuit` object.
+            coupling_map: The reduced coupling map of the backend. For example,
+                if physical qubits [5, 6, 7] to be benchmarked is connected
+                as 5 - 7 - 6 linearly, the reduced coupling map is 0 - 2 - 1.
+            target: A target representing the target backend, which will be ignored in this plugin.
+            qubits: List of physical qubits over which the operation is defined,
+                which will be ignored in this plugin.
+            options: Additional method-specific optional kwargs,
+                which must include ``basis_gates``, basis gates to be used for the synthesis.
+
+        Returns:
+            The quantum circuit representation of the Operation
+                when successful, and ``None`` otherwise.
+
+        Raises:
+            QiskitError: If basis_gates is not supplied.
+        """
+        # synthesize cliffords
+        circ = synth_clifford_full(high_level_object)
```

Comment:
I’ve examined `synth_clifford_depth_lnn` using random RB circuits on 3 linear qubits (i.e. no need for SabreSwap routing) and found that it is faster as expected but surprisingly it produces circuits with more CNOT counts and deeper depths than `synth_clifford_full` + SabreSwap. So I’ll stick to `synth_clifford_full` in the PR.

Reply:
synth_clifford_depth_lnn:
```
total cx counts = 2961, depth = 4418, max_length=10, time_transpiled_circuits:	1.233635034004692
total cx counts = 6286, depth = 9193, max_length=20, time_transpiled_circuits:	1.9450378390029073
total cx counts = 8659, depth = 12850, max_length=30, time_transpiled_circuits:	2.565633656005957
total cx counts = 11074, depth = 16313, max_length=40, time_transpiled_circuits:	3.270593570006895
total cx counts = 13727, depth = 20190, max_length=50, time_transpiled_circuits:	3.964288625997142
```

synth_clifford_full:
```
total cx counts = 1589, depth = 3092, max_length=10, time_transpiled_circuits:	1.8188774910086067
total cx counts = 2915, depth = 6009, max_length=20, time_transpiled_circuits:	2.883968324007583
total cx counts = 4120, depth = 8318, max_length=30, time_transpiled_circuits:	3.982568758990965
total cx counts = 5439, depth = 10996, max_length=40, time_transpiled_circuits:	4.483078286997625
total cx counts = 6493, depth = 13110, max_length=50, time_transpiled_circuits:	5.633502009004587
```